adding regularization blurb

jaybee84 opened the pull request:
### Purpose

Adding a section on regularization 

### Directions for reviewers

The writing is still a bit rough and the Deep learning regularization tricks part is still not finished. Focus is requested on the content and any additions that need to be done.

#### Which areas should receive a particularly close look?

Content

#### Is there anything that you want to discuss further?


#### Is the pull request ready for review?
Yes

### Manuscript checklist

<!--Remove this section if the changes are not relevant to the manuscript itself-->

Unless otherwise noted above, this PR will be considered ready for review when all four items have been checked.

- [ X] All changes to text follow "one sentence per line" [[Manubot instructions](https://github.com/jaybee84/ml-in-rd/blob/draft-branch/USAGE.md#manubot-markdown)]
- [ X] All citations follow the [Manubot citation instructions](https://github.com/jaybee84/ml-in-rd/blob/draft-branch/USAGE.md#citations)
- [ ] All changes or additions to tables follow the [Manubot table instructions](https://github.com/jaybee84/ml-in-rd/blob/draft-branch/USAGE.md#tables)
- [ ] All figures follow the [Manubot figure instructions](https://github.com/jaybee84/ml-in-rd/blob/draft-branch/USAGE.md#figures)

jaybee84 committed:
adding regularization blurb

jaybee84 committed:
making one sentence one line

allaway commented on content/04.model-complexity.regularization.md:
```diff
@@ -0,0 +1,55 @@
+
+Machine learning algorithms are optimized to find patterns among data points. 
+Given a large amount of data, this optimization works in favor of the algorithm trying to learn complex patterns in the data. 
+However, when data is sparse, the same optimization leads to the model learning technical variances rather than actual patterns in the data. 
```
One or two example citations to support this might be helpful!

allaway replied:
I also think that  "sparse" can mean different things to different readers. In my experience rare disease datasets are small, but not sparse in the sense that they have lots of zero values. This might be a good place to talk more in depth about p>>n problem, and clarify what sparse is referring to here.

jaclyn-taroni replied:
Agree about expecting to see a citation here

allaway commented on content/04.model-complexity.regularization.md:
```diff
@@ -0,0 +1,55 @@
+However, when data is sparse, the same optimization leads to the model learning technical variances rather than actual patterns in the data. 
+This results in low bias but high variance in the prediction results i.e. overfitted models fit the training data very closely (low bias), but fail to predict new testing data with high accuracy (high variance) leading to low prediction accuracy of the model in view of new data. 
+Such overfitting can lead to misinterpretation of technical variances as true patterns in the sparse data and thus needs to be minimized. 
+Minimization of overfitting can be accomplished by crossvalidation and regularization methodologies. 
```
I tend to see it written as "cross-validation" rather than one word.

allaway commented on content/04.model-complexity.regularization.md:
```diff
@@ -0,0 +1,55 @@
+Such overfitting can lead to misinterpretation of technical variances as true patterns in the sparse data and thus needs to be minimized. 
+Minimization of overfitting can be accomplished by crossvalidation and regularization methodologies. 
+While crossvalidation aims to reduce the variance in prediction, regularization adds a small amount of bias to the initial model to minimize its dependance and sensitivity to training data. 
+Regularization makes models less sensitive to variance in sample data by adding a penalty (lambda) to the sum of squared residuals (or sum of maximum likelihoods). 
```
This couple of sentences might be a bit hard for a novice-to-ML reader to follow without a figure, or at least citation describing this in more detail. I don't think it's the role of this perspective to teach these concepts outright, but perhaps this could be written in a way that it could be accessible to a less expert reader.

jaclyn-taroni replied:
I agree. There are definitely _practical considerations_ for all of what is included here. We could approach making things more digestible through some concrete examples. 

The concrete example idea is not specific to this section, but an approach we might want to discuss more generally.

jaybee84 replied:
reframed and simplified this part in the new commit

allaway commented on content/04.model-complexity.regularization.md:
```diff
@@ -0,0 +1,55 @@
+
+To learn from 'n' number of features, machine learning models ideally need at least 'n+1' samples. 
+For eg. to predict any disease outcome from 20,000 genes we need at least 20,001 samples. 
+However, in reality even the best sample set probably contains 500 samples, and in case of rare diseases that may be closer to 50 than 500. 
```
A citation to support this would be helpful

jaybee84 replied:
edited this part out, since it was mainly a conceptual approximation without any specific citation.

allaway committed:
Update 04.model-complexity.regularization.md

allaway commented on content/04.model-complexity.regularization.md:
```diff
@@ -0,0 +1,55 @@
+However, in reality even the best sample set probably contains 500 samples, and in case of rare diseases that may be closer to 50 than 500. 
+Regularization can not only minimize overfitting but can additionally help in predicting outcomes using fewer number of samples. 
+Regularization can be of three main types, each with their particular strengths and weaknesses. 
+(1) Ridge regression aims to minimize the magnitude of the features, but in models that try to select the most important features for accurate prediction of sample labels, ridge regression shrinks all features equally, but can never shrink features all the way to 0. 
```
Changed "ridge" and "lasso" to make them more consistent.

allaway replied:
Would suggest changing "shrink features all the way to 0" to something conceptually clearer like "reduce the number of features" or "remove unimportant features entirely"

allaway commented on content/04.model-complexity.regularization.md:
```diff
@@ -0,0 +1,55 @@
+(1) Ridge regression aims to minimize the magnitude of the features, but in models that try to select the most important features for accurate prediction of sample labels, ridge regression shrinks all features equally, but can never shrink features all the way to 0. 
+Thus in presence of many correlated parameters (eg. gene expression networks), ridge regression may not be ideal in reducing the feature space. 
+(2) LASSO regression on the other hand works well for selecting few important features since its effect can shrink some features more than the others. 
+Thus it helps in selecting most important features while irrelevant features are lost and shrunk to 0. This selection attribute of LASSO (in a sample set of size "n", LASSO can only select "n" features for the model) may be an advantage in reducing model complexity, but a disadvantage in cases where identification of all possible collinear features is important (eg. all biomarkers correlating to  ...
```
needs a new line for second sentence :)

jaclyn-taroni replied:
Like the inclusion of `eg. all biomarkers correlating to a particular disease phenotype`!

allaway committed:
Update 04.model-complexity.regularization.md

allaway commented on content/04.model-complexity.regularization.md:
```diff
@@ -0,0 +1,55 @@
+(3) Elastic net regression is a combination of LASSO and ridge regression[@doi:10.1111/j.1467-9868.2005.00503.x]. 
+Both of the methodologies when applied together helps to select most useful features, specially where there are a lot of correlated features. 
+In this setup, LASSO leads to selection of one of the correlated features and reduces the others to 0 (grouping of features). 
+The ridge regression aspect of elastic-net then minimizes the magnitude of the features that were selected without shrinking any of them to 0 (minimization). 
```
To the lay-ish reader, the value of ridge here might not be so clear.

jaybee84 replied:
simplified this in the new commit

allaway commented on content/04.model-complexity.regularization.md:
```diff
@@ -0,0 +1,55 @@
+While ridge regression is not usually utilized for feature selection, adaptive ridge regression has been utilized to help combine rare variants into a single score analogous to feature engineering for increasing the signal of rare variants[@doi:10.1371/journal.pone.0044173]. 
+Another variation of LASSO included its integration with the probabilistic logistic bayesian approach to identify a protective rare variant in lung cancer[@doi:10.4137/CIN.S17290]. 
+Xu et al. on the other hand combined the feature selection methods with a generalized pooling strategy, and evaluated the performance of these hybrid approaches for detection of rare genetic variants[@doi:10.1371/journal.pone.0041694]. 
+Another interesting approach is the sparse-group LASSO approach which incorporates prior knowledge into the regularization[@doi:10.1080/10618600.2012.681250]. 
```
I wonder if this should be here or the "prior knowledge" section

jaybee84 replied:
thats a good point, will keep it here for now... but may need to reposition when we put the sections together.

allaway commented on content/04.model-complexity.regularization.md:
```diff
@@ -0,0 +1,55 @@
+Another interesting approach is the sparse-group LASSO approach which incorporates prior knowledge into the regularization[@doi:10.1080/10618600.2012.681250]. 
+This approach works well for a scenario where only few genes in a pathway are true predictors of a phenotype, where it helps select the driving genes in a pathway of interest. 
+Alternatively, a combination of LASSO and ridge regression, Elastic-net regression has also been used to reduce the feature space in various types of cancer datasets [@doi:10.1016/j.compbiomed.2015.10.008; @doi:10.1186/1471-2105-14-198]. 
+In cases where the number of features were far greater than the number of samples, it has usually been found to outperform the other regression approaches [@doi:10.1111/j.1467-9868.2005.00503.x]. 
```
is "it" an ensemble of these three methods?

jaybee84 replied:
"it" here refers to elastic-net.. will reframe the sentence accordingly to avoid confusion

allaway reviewed (approved):
Looks like a great start! I suggested some changes, but think that this could be merged as-is since it is the initial commit for this section.

jaclyn-taroni commented on content/04.model-complexity.regularization.md:
```diff
@@ -0,0 +1,55 @@
+Machine learning algorithms are optimized to find patterns among data points. 
+Given a large amount of data, this optimization works in favor of the algorithm trying to learn complex patterns in the data. 
+However, when data is sparse, the same optimization leads to the model learning technical variances rather than actual patterns in the data. 
+This results in low bias but high variance in the prediction results i.e. overfitted models fit the training data very closely (low bias), but fail to predict new testing data with high accuracy (high variance) leading to low prediction accuracy of the model in view of new data. 
```
```suggestion
This results in low bias but high variance in the prediction results, i.e., overfitted models fit the training data very closely (low bias), but fail to predict new testing data with high accuracy (high variance) leading to low prediction accuracy of the model in view of new data. 
```

jaclyn-taroni commented on content/04.model-complexity.regularization.md:
```diff
@@ -0,0 +1,55 @@
+In this setup, LASSO leads to selection of one of the correlated features and reduces the others to 0 (grouping of features). 
+The ridge regression aspect of elastic-net then minimizes the magnitude of the features that were selected without shrinking any of them to 0 (minimization). 
+
+Any supervised learning implementation in rare disease would require robustness towards feature selection from a small number of samples. 
```
Continuing with trying to make this a bit "friendlier" to someone new to ML, I think we could be more specific about what we mean by _robustness_ (e.g., what is the practical implication). Particularly because the next sentence begins with "This robustness."

jaclyn-taroni commented on content/04.model-complexity.regularization.md:
```diff
@@ -0,0 +1,55 @@
+
+Any supervised learning implementation in rare disease would require robustness towards feature selection from a small number of samples. 
+This robustness is mostly acquired through the combination of various regression strategies. 
+Since machine learning applications in rare disease is infrequent, combination strategies used for rare variant discovery and immune cell signature discovery can serve as good case studies to examine. 
```
```suggestion
Since machine learning applications in rare disease are infrequent, combination strategies used for rare variant discovery and immune cell signature discovery can serve as good case studies to examine. 
```

jaclyn-taroni commented on content/04.model-complexity.regularization.md:
```diff
@@ -0,0 +1,55 @@
+This approach works well for a scenario where only few genes in a pathway are true predictors of a phenotype, where it helps select the driving genes in a pathway of interest. 
+Alternatively, a combination of LASSO and ridge regression, Elastic-net regression has also been used to reduce the feature space in various types of cancer datasets [@doi:10.1016/j.compbiomed.2015.10.008; @doi:10.1186/1471-2105-14-198]. 
+In cases where the number of features were far greater than the number of samples, it has usually been found to outperform the other regression approaches [@doi:10.1111/j.1467-9868.2005.00503.x]. 
+A variation of the elastic net regression was used for identifying immune cell signatures in an RNASeq dataset where the number of cells sampled were far fewer than number of genes profiled [@doi: 10.1186/s12859-019-2994-z]. 
```
```suggestion
A variation of the elastic net regression was used for identifying immune cell signatures in an RNA-seq dataset where the number of cells sampled were far fewer than number of genes profiled [@doi: 10.1186/s12859-019-2994-z]. 
```
Is usually what I use but I think this is one of those scenarios where there are multiple correct ways to stylize something and we just have to be consistent!

jaclyn-taroni commented on content/04.model-complexity.regularization.md:
```diff
@@ -0,0 +1,55 @@
+For its feature selection attribute, LASSO or least absolute shrinkage and selection operator has been widely applied in microarray and GWAS data for common variants. 
+But since LASSO by itself is too stringent for rare variants, it has been employed along with group penalties to help identify rare variants/ low frequency predictors [@doi:10.1093/bioinformatics/btq448]. 
+Variations of LASSO have also been implemented to aggregate or group the occurrance of rare variants together by gene or chromosome location [@doi:10.1002/gepi.21746; @doi:10.1186/1753-6561-5-S9-S100; @doi:10.1016/j.ajhg.2008.06.024]. 
+In this strategy, a 0–1 dummy variable was created for each SNP based on the presence or absence of the rare variant. 
+Then linear combinations of the selected dummy variables were considered by using the LASSO procedure. 
+Even though most of the dummy variables were 0, their linear combination was more likely to be nonzero thus leading to increased signal to noise ratio for the rare variants. 
```
I'm not sure if everyone in the intended audience will know the term dummy variable (possibly related to #24 depending on the scope of that "box").

jaybee84 replied:
added to issue #50

jaclyn-taroni reviewed (commented):
Thanks again for getting us started @jaybee84! @allaway left some comments about making this section more digestible for new practitioners. In service of this goal, I think it'd be helpful to discuss the idea of including some concrete/practical examples when introducing the concepts like bias, variance, etc. (perhaps during our check-in this afternoon!) You've included specific examples in the form of particular studies as the last part of the section. I'm wondering if we can have some more abstract examples in the beginning of the section. They'd need to not be so abstract as to no longer be appropriate for an audience of new practitioners, so could be a little tricky. Looking forward to everybody's thoughts!

jaybee84 committed:
Apply suggestions from review - part1

Co-authored-by: Jaclyn Taroni <jaclyn.n.taroni@gmail.com>

jaybee84 committed:
adding suggestions from review -part1

jaybee84 committed:
Merge branch 'jaybee84/10-regularization' of https://github.com/jaybee84/ml-in-rd into jaybee84/10-regularization

jaybee84 committed:
update part1 commit

cgreene commented on content/04.model-complexity.regularization.md:
```diff
@@ -0,0 +1,57 @@
+While the value of lambda is determined by crossvalidation, the minimization of the sum of squared residuals (or error) leads to a model that is less sensitive to changes in sample data.
+
+To learn from 'n' number of features, machine learning models ideally need at least 'n+1' samples. 
+For eg. to predict any disease outcome from 20,000 genes we need at least 20,001 samples. 
```
```suggestion
For example, to predict any disease outcome from 20,000 genes we need at least 20,001 samples. 
```

jaybee84 committed:
Update content/04.model-complexity.regularization.md

Co-authored-by: Casey Greene <cgreene@users.noreply.github.com>

cgreene commented:
Hmm - GitHub usually offers "update to master" as a button. However, with how we're using a branching model that isn't happening here. Without updating to the current draft-branch before merging, you won't get spellchecking. Once we start making PRs off a draft-branch version that does have spellchecking included, we'll get it.

jaybee84 commented:
sounds good.

jaybee84 committed:
incorporating suggestion from earlier comments
1. adds references where needed
2. adds example in the beginning to simplify concepts

jaybee84 committed:
Merge branch 'jaybee84/10-regularization' of https://github.com/jaybee84/ml-in-rd into jaybee84/10-regularization

jaybee84 commented:
closes #10
